Sources/MongoSwift/BSON/BSONValue.swift

```diff
@@ -524,12 +675,39 @@ extension Double: BSONValue {
 }
 
 /// An extension of `Int` to represent the BSON Int32 or Int64 type.
-/// The `Int` will be encoded as an Int32 if possible, or an Int64 if necessary.
-extension Int: BSONValue {
-    public var bsonType: BSONType { return self.int32Value != nil ? .int32 : .int64 }
+/// On 64-bit systems, `Int` corresponds to a BSON Int64. On 32-bit systems, it corresponds to a BSON Int32.
+extension Int: BSONNumber {
+    /// `Int` corresponds to a BSON int32 or int64 depending upon whether the compilation system is 32 or 64 bit.
+    /// Use MemoryLayout instead of Int.bitWidth to avoid a compiler warning.
+    /// See: https://forums.swift.org/t/how-can-i-condition-on-the-size-of-int/9080/4
+    internal static var bsonType: BSONType {
+        return MemoryLayout<Int>.size == 4 ? .int32 : .int64
+    }
+
+    public var bsonType: BSONType { return Int.bsonType }
+
+    internal var int32Value: Int32? {
+        guard self.bsonType == .int32 else {
+            return nil
+        }
+        return Int32(exactly: self)
+    }
+
+    internal var int64Value: Int64? {
+        guard self.bsonType == .int64 else {
+            return nil
+        }
+        return Int64(exactly: self)
+    }
 
-    internal var int32Value: Int32? { return Int32(exactly: self) }
-    internal var int64Value: Int64? { return Int64(exactly: self) }
+    /// Initialize this `Int` from another `BSONNumber`.
+    /// If the passed in number cannot be represented exactly as an `Int`, this initializer will return nil.
+    public init?(fromNumber num: BSONNumber) {
+        guard let int = num.toInt() else {
+            return nil
+        }
+        self = int
+    }
 
     public func encode(to storage: DocumentStorage, forKey key: String) throws {
         if let int32 = self.int32Value {
```

Comment:
I think it might be safer if we explicitly switch on the bson type here, rather than relying on the properties to check it for us.

Reply:
The thought process here is that the `nil` check constitutes the bson type check, and that we'd still have to perform the nil check after performing the bson type check even though we'd know it would always succeed. Given that, do you still think it's worth the explicit type check?

Reply:
hm, I guess not. maybe this is just a naming thing... I feel like `int32Value` doesn't convey that we are doing the type check, it kind of sounds like it's just doing the same thing as `toInt32()`.  no immediate ideas for a better name though...

Reply:
We could make it simply `int32` and `int64`. Then if `self.int32` is nil, it slightly implies that `self` isn't an `int32`. 

Similarly, `asInt32` and `asInt64` imply cast-like behavior, which is close to what is going on here. If they fail, then `Int` isn't whatever was being "casted" to.

A third option is to provide only a single accessor that returns whatever it actually is (typed as a `BSONNumber`), which can then itself be casted. The property could be called `typedValue` or something along those lines.
e.g. 
```swift
if let int32 = self.typedValue as? Int32 { ... }
else if let int64 = self.typedValue as? Int64 { ... }

switch self.typedValue {
case let int32 as Int32:
    // do stuff with int32
case let int64 as Int64:
    // do stuff with int64
default:
    // should be impossible
}
```

Reply:
As part of updating the protocol to use computed properties, I had to rename `int32Value` and `int64Value` so they wouldn't collide with the protocol. I renamed them to just `int32` and `int64` arbitrarily, and am still open to other rename candidates.

Related to naming collisions: I think `NSNumber` won't be able to conform to `BSONNumber` because its computed properties don't return optionals, but ours do. I don't think this is the biggest problem in the world because `NSNumber` is basically a vehicle for achieving the same stuff as `BSONNumber`, just with the focus on C types instead of BSON types, so I don't think there's any meaningful value in conforming it to `BSONValue`. Honestly, there's not a ton of value in conforming _anything_ to `BSONValue` outside of the types we've done it for since we don't expose any way of writing to the `bson_t`.

Reply:
hmmm..... I kinda like the `typedValue` option just because it's name is the clearest. 

also, since it returns a `BSONNumber` which inherits from `BSONValue`, in a lot of these places it would actually eliminate the need to handle them separately altogether -- for example
```swift
if let int32 = self.int32 {
    return try int32.encode(to: storage, forKey: key)
}
if let int64 = self.int64 {
    return try int32.encode(to: storage, forKey: key)
}
```
could just become: `return try typedValue.encode(to: storage, forKey: key)`.

Reply:
(or maybe you would need to unwrap `typedValue` first, but anyway, saves you the two separate `encode` calls.)

Reply:
Since we know `Int` will fit in either `Int32` or `Int64`, we can just return a regular `BSONNumber` and won't need to unwrap.

Reply:
(also, I've updated to use the `typedValue` computed property)